maven-plugin/src/test/java/io/confluent/kafka/schemaregistry/maven/derive/schema/DeriveJsonSchemaTest.java

```diff
@@ -0,0 +1,132 @@
+    generateSchemaAndCheckExpected(complexTypesWithPrimitiveValues, expectedSchema);
+  }
+
+  @Test
+  public void testDeriveRecordComplexTypesRecursive() throws IOException {
+    // Get schema for record with arrays and records having complex types
+    String recordOfArrays = "\"RecordOfArrays\": {\"ArrayInt1\": [12], \"ArrayEmpty\": []}";
+    String recordOfRecords = "\"RecordOfRecords\": {\"Record1\": {\"name\": \"Tom\"}, \"Record2\": {\"place\": \"Bom\"}}";
+    String complexTypesWithPrimitiveValues = "{" + recordOfArrays + "," + recordOfRecords + "}";
+    String expectedSchema = "{\"type\":\"object\",\"properties\":{\"RecordOfArrays\":{\"type\":\"object\",\"properties\":{\"ArrayEmpty\":{\"type\":\"array\",\"items\":{}},\"ArrayInt1\":{\"type\":\"array\",\"items\":{\"type\":\"number\"}}}},\"RecordOfRecords\":{\"type\":\"object\",\"properties\":{\"Record1\":{\"type\":\"object\",\"properties\":{\"name\":{\"type\":\"string\"}}},\"Record2\":{\"type\":\"object\",\"properties\":{\"place\":{\"type\":\"string\"}}}}}}}";
+    generateSchemaAndCheckExpected(complexTypesWithPrimitiveValues, expectedSchema);
+  }
+
+  @Test
+  public void testDeriveRecordWithArrayOfDifferentTypes() throws IOException {
+    // Array has elements of type number, float and string represented as oneOf in schema
+    String arrayOfDifferentTypes = "{\"ArrayOfDifferentTypes\": [2, 13.1, true, \"J\", \"K\"]}";
+    String expectedSchema = "{\"type\":\"object\",\"properties\":{\"ArrayOfDifferentTypes\":{\"type\":\"array\",\"items\":{\"oneOf\":[{\"type\":\"boolean\"},{\"type\":\"number\"},{\"type\":\"string\"}]}}}}";
+    generateSchemaAndCheckExpected(arrayOfDifferentTypes, expectedSchema);
+
+    // Array of Records with arrays and different records, checking recursive merging of records
+    String arrayOfRecordsAndArrays = "{\"ArrayOfRecordsAndArrays\": [ {\"J\":[1,11]}, {\"J\":{\"J\":12}},  {\"J\":{\"J\": true}}]}";
+    String expectedSchema2 = "{\"type\":\"object\",\"properties\":{\"ArrayOfRecordsAndArrays\":{\"type\":\"array\",\"items\":{\"type\":\"object\",\"properties\":{\"J\":{\"oneOf\":[{\"type\":\"array\",\"items\":{\"type\":\"number\"}},{\"type\":\"object\",\"properties\":{\"J\":{\"oneOf\":[{\"type\":\"boolean\"},{\"type\":\"number\"}]}}}]}}}}}}";
+    generateSchemaAndCheckExpected(arrayOfRecordsAndArrays, expectedSchema2);
+  }
+
+  @Test
+  public void testDeriveMultipleMessages() throws JsonProcessingException {
+    // Merging Records with different field names
+    String arrayOfStrings = "{\"ArrayString\": \"1\"}";
+    String arrayOfIntegers = "{\"ArrayInteger\": 12}";
+    String expectedSchema = "{\"schema\":{\"type\":\"object\",\"properties\":{\"ArrayInteger\":{\"type\":\"number\"},\"ArrayString\":{\"type\":\"string\"}}}}";
+    ObjectNode schema = derive.getSchemaForMultipleMessages(Arrays.asList(arrayOfStrings, arrayOfIntegers, arrayOfStrings, arrayOfIntegers));
+    assertEquals(expectedSchema, schema.toString());
+  }
+}
```

Comment:
ideally there should be a test for the static function as well `sortJsonArrayList`